src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -2604,16 +2618,28 @@ private void GenerateTypesAndUsings(ScriptBlockAst rootForDefiningTypesAndUsings
             {
                 if (rootForDefiningTypesAndUsings.UsingStatements.Count > 0)
                 {
-                    bool allUsingsAreNamespaces = rootForDefiningTypesAndUsings.UsingStatements.All(us => us.UsingStatementKind == UsingStatementKind.Namespace);
+                    bool allUsingsAreNamespaces = true;
+                    foreach (UsingStatementAst @using in rootForDefiningTypesAndUsings.UsingStatements)
+                    {
+                        if (@using.UsingStatementKind != UsingStatementKind.Namespace)
+                        {
+                            allUsingsAreNamespaces = false;
+                            break;
+                        }
+                    }
+
                     GenerateLoadUsings(rootForDefiningTypesAndUsings.UsingStatements, allUsingsAreNamespaces, exprs);
                 }
+                var typeAsts = new List<TypeDefinitionAst>();
+                var astList = rootForDefiningTypesAndUsings.FindAll(
+                    ast => ast is TypeDefinitionAst,
+                    searchNestedScriptBlocks: true);
+                foreach (Ast ast in astList)
+                {
+                    typeAsts.Add((TypeDefinitionAst)ast);
+                }
```

Comment:
It's a shame that the return type for `FindAll` is `IEnumerable<>` or you'd be able to do `new TypeDefinitionAst[astList.Length]` and avoid the dynamic array building from `List<>`.

Maybe out of scope of this PR, but it'd be nice to see some (or all) of these `FindAll`'s replaced with `AstVisitor`'s. e.g.

```csharp
private sealed class TypeDefFinder : AstVisitor2
{
    private List<TypeDefinitionAst> _found;

    public static TypeDefinitionAst[] FindAll(Ast root)
    {
        var visitor = new TypeDefFinder();
        root.Visit(visitor);
        return visitor._found?.ToArray() ?? Array.Empty<TypeDefinitionAst>();
    }

    public override AstVisitorAction VisitTypeDefinition(TypeDefinitionAst typeDefinitionAst)
    {
        (_found ??= new()).Add(typeDefinitionAst);
        return AstVisitAction.SkipChildren;
    }
}
```

Reply:
I had the same "ah, that returns IEnumerable.... *shakes head*" 😂 

I'm more than happy to do that as a follow up tbh. 🙂

Reply:
Current implementation of the FindAll() returns List<>. So we need only change the method return type.

Reply:
`FindAll()` is a public method, I don't think we can do that at this point in time 😅

Reply:
```
internal static IEnumerable<Ast> FindAll(Ast ast, Func<Ast, bool> predicate, bool searchNestedScriptBlocks)
```
It is internal.

Reply:
Not sure which method/overload you're looking at there, but my IDE tells me this calls this particular overload of FindAll()
![image](https://user-images.githubusercontent.com/32407840/116948795-70831f00-ac4e-11eb-8e97-d83bee5683fd.png)

(ast.cs, line 150)

Reply:
I suggest to update and use the internal method I pointed. 
Also please use `static` modifier for predicate.

Reply:
I'll check it out this evening, thanks for the reminder! 💖